app/src/main/java/com/toandv/demo/musicplayer/services/MusicPlayer.kt

```diff
@@ -0,0 +1,146 @@
+                it.start()
+                listener.onPlayingChanged(true)
+                startUpdate()
+            }
+
+            setOnCompletionListener {
+                isPrepared = false
+                listener.onPlayingChanged(false)
+                if (!it.isLooping) next()
+                stopUpdate()
+            }
+
+            setOnErrorListener { _, what, extra ->
+                listener.onError("Error($what, $extra)")
+                stopUpdate()
+                false
+            }
+        }
+    }
+
+    fun updatePlayList(list: List<Song>) {
+        playList.clear()
+        playList.addAll(list)
+        if (pos == -1 && playList.isNotEmpty()) pos = 0
+    }
+
+    fun setSong(id: Long): Boolean {
+        pos = playList.indexOfFirst {
+            it.id == id
+        }
+        return when (pos) {
+            -1 -> false
+            else -> {
+                playCurrentSong()
+                true
+            }
+        }
+    }
+
+    fun playOrPause() {
+        when {
+            player.isPlaying -> {
+                player.pause()
+                listener.onPlayingChanged(false)
+                stopUpdate()
+            }
+            isReady && player.isPlaying.not() -> when {
+                isPrepared -> {
+                    player.start()
+                    listener.onPlayingChanged(true)
+                    startUpdate()
+                }
+                else -> playCurrentSong()
+            }
+        }
+
+    }
+
+    fun loop() {
+        player.isLooping = !player.isLooping
+    }
+
+    fun next() {
+        when {
+            isReady.not() -> return
+            pos < playList.lastIndex -> pos++
+            else -> pos = 0
+        }
+        playCurrentSong()
+    }
+
+    fun previous() {
+        when {
+            isReady.not() -> return
+            pos == 0 -> pos = playList.lastIndex
+            else -> pos--
+        }
+        playCurrentSong()
+    }
+
+    fun seekTo(newPos: Int) {
+        if (isReady && isPrepared) player.seekTo(newPos)
+    }
+
+    fun restoreDetail() {
+        if (isReady && isPrepared) {
+            listener.onSongPrepared(playList[pos], player.duration)
+            listener.onPlayingChanged(isPlaying)
+            if (isPlaying) listener.onSeekChanged(player.currentPosition)
+        }
+    }
+
+    val isPlaying get() = player.isPlaying
+
+    private val isReady get() = playList.isNotEmpty() && pos != -1
+
+    private fun startUpdate() {
+        timer = Timer()
+        timer.schedule(object : TimerTask() {
+            override fun run() {
+                if (player.isPlaying) listener.onSeekChanged(player.currentPosition)
+            }
+        }, 0, 1000)
+    }
+
+    private fun stopUpdate() {
+        timer.cancel()
+        timer.purge()
+    }
+
+    private fun playCurrentSong() {
+        player.reset()
+        player.setDataSource(context, Uri.parse(playList[pos].source))
+        player.prepareAsync()
```

Comment:
Viết theo cách của Kotlin nhé 
```suggestion
        with(player){
            reset()
            setDataSource(context, Uri.parse(playList[pos].source))
            prepareAsync()
        }     
```